README.markdown
# Py65 for Python 3

Note: This is a Python 3 port of Mike Naberezny's Py65 project
performed by David Beazley (http://www.dabeaz.com) for his own
diabolical amusement and personal projects.  Important differences
from the standard version:

- Py65 is only installed as a library.  If you want to run the
  monitor, use 'python3 -m py65.monitor'.

- Installation is based on distutils, not setuptools.

- Installation should not conflict with an existing Python-2
  installation of py65.

- Includes a 'superboard' demo where you can run an emulated
  1979 OSI Superboard II system and write little Microsoft 
  Basic 1.0 programs.   It's fun.

So far as I know, this version passes all of the tests contained 
in src/py65/tests. However, I have not done any testing beyond
my own use of the library (e.g., "it works for me.")

Note: I do not intend to maintain this branch except as needed for my
own projects under Python 3.  Use at your own risk. - Dave.

# Py65

Py65 provides tools for simulating hardware based on 6502-like
microprocessors.  It has the following goals:

 - Focus on ease of use and modularity rather than performance.  Py65 is
   written in the Python programming language for productivity, while
   similar programs are written in C for performance.

 - Enable simulations to be created for systems where it might have 
   otherwise not been practical, such as homebuilt computers. 

 - Rigorously unit test all of the components.  While the tools provided
   by Py65 may not always be perfect, their behavior is verified through 
   tests so unexpected results are minimized.
   
## Installation

Obtain the source from Github (http://github.com/dabeaz/py65).  Please
note that the original location of the Py65 project is at
http://github.com/mnaberez/py65/downloads.   Use the following to install
after you've obtained the source:

    $ python3 setup.py install

## Devices

The following devices are simulated at this time:

 - `mpu6502` simulates the original NMOS 6502 microprocessor from MOS
    Technology, later known as Commodore Semiconductor Group (CSG). At this
    time, all of the documented opcodes are supported.  Support for the
    illegal opcodes is planned for the future.

 - `mpu65c02` simulates a generic CMOS 65C02 microprocessor. There were
    several 65C02 versions from various manufacturers, some with more opcodes
    than others. This simulation is based on the W65C02S from the Western
    Design Center (WDC).

## Monitor

Py65 includes a console-based machine language monitor (sometimes also
called a debugger).  This program, launched using 'python3 -m
py65.monitor' allows you to interact with the simulations that you
build.  Its features include:

 - Commands that are largely compatible with those used in the monitor of
   the popular VICE emulator for Commodore computers.

 - Ability to load, dump, and fill memory.

 - Simple assemble and disassemble capability, including support for labels 
   and labels with offsets.

## Documentation

Py65 documentation is written using [Sphinx](http://sphinx.pocoo.org/) and is
periodically published to 
[http://6502.org/projects/py65/](http://6502.org/projects/py65/).

## Contributors

These people are responsible for Py65:

 - [Mike Naberezny](http://github.com/mnaberez) is the original author of 
   Py65 and is the primary maintainer.
 
 - [Oscar Lindberg](http://github.com/offe) started the 65C02 simulation 
   module and contributed greatly to its implementation. 

 - [Ed Spittles](http://github.com/biged) helped with testing and provided 
   many useful issue reports and patches.

 - [Dave Beazley](http://github.com/dabeaz) is the guilty party responsible
   for this Python 3 port.


CHANGES.txt
0.8-py3k

  - Port of version 0.8 to Python 3 by David Beazley. 

0.8 (2010-03-08)

  - Fixed deprecation warnings on Python 2.6

  - We no longer bundle ez_setup to bootstrap setuptools installation.

  - Restoring the processor status register from interrupt now correctly
    set the BREAK and UNUSED flags to be high.  Thanks to Ed Spittles
    for reporting this.

  - Applied patch by Ed Spittles that fixes the behavior of the BREAK
    and UNUSED flags in the processor status register.  Closes #16.

  - Added ">" as a monitor shortcut for the fill command for 
    consistency with VICE.
  
0.7 (2009-09-03)

  - When using the monitor, the nonblocking character input at
    $F004 should now work on the Microsoft Windows platform.

  - Fixed that relative branch calculations would not use the correct 
    start address when assembling in the monitor.  Closes #10.

  - The processor status register ("p" or "flags") can now be changed 
    in the monitor using the "registers" command with an argument of
    "p", such as "registers p=00".

  - MPU objects now return a two-line string as their __repr__ with 
    the processor status register displayed as binary for readability.  

  - The processor status register is now initialized to 0 on reset.
    Previously, its unused bit (bit 5) was set to 1 on reset.

  - Applied patch from Ed Spittles to change the CMP algorithm so that
    it no longer fails Rob Finch's test suite.  Closes #8. 

  - Added a new interactive assembly mode to the monitor.  Entering the
    the assemble command with a statement such as "a c000 lda #0" works
    as before.  Entering "a c000" will start the interactive assembler
    at that address.  Entering "a" alone will start it at the current
    program counter.

  - Applied patch from Ed Spittles so that SBC now properly sets the
    Overflow (V) flag.  This fixes a failure in Rob Finch's test suite.
    Closes #6.

  - Applied patch from Ed Spittles so that SBC now properly sets the
    Carry (C) and Zero (Z) flags.  This fixes failures caught by Ed's
    own tests (see http://forum.6502.org/viewtopic.php?p=8854#8854).
    Closes #15.

  - A new "save" command has been added to the monitor that will save
    a range of memory to a binary file.

0.6 (2009-08-11)

  - Added monitor shortcut "a" for "assemble".

  - Fixed that ASL would not properly set the Z flag.  Closes #7.

  - Fixed that ADC would not properly set the Overflow (V) flag.  The
    overflow calculation that is now used originated from XGS: Apple
    IIGS Emulator (cputable.h).  Originally written and Copyright 
    (C)1996 by Joshua M. Thompson.  Copyright (C) 2006 by Samuel A. 
    Falvo II.  http://bitbucket.org/kc5tja/lib65816/src/tip/src/cputable.h
    Closes #3.

0.5 (2009-08-06)

  - Fixed signatures of getc/putc callbacks in monitor that were broken
    when the ObservableMemory interface changed in 0.3.  Closes #1.

  - Fixed that ROL would not properly set the Z flag.  Closes #2.

0.4 (2009-06-06)

  - Added ez_setup.py to bootstrap setuptools installation.

0.3 (2009-06-03)

  - Added shortcuts for monitor commands such as "m" for "memory".  These
    are mostly the same as the VICE monitor shortcuts.

  - The terminal width can now be changed in the monitor using the new
    "width" command.  Some commands, like "mem", will wrap to this width.

  - Fixed a bug where BRK would increment PC by 3 instead of 2.  Thanks
    to Oscar Lindberg.

  - Added a new 65C02 MPU simulation started by Oscar Lindberg.  It is
    now mostly complete.
    
  - Added a new "mpu" command to the monitor.  It will switch between the
    NMOS 6502 and CMOS 65C02 simulations.

  - A new "devices" module has been added to organize device simulations.

  - The mpu6502 and mpu65c02 devices have been reorganized internally to
    use Python decorators to build their lookup tables based on an
    idea by Oscar Lindberg.

  - A new "utils" module has been added with various utility functions.

  - The ObservableMemory interface has been changed for clarity.

  - Python 2.4 or later is now required.

0.2 (2008-11-09)

  - Added a new "disassemble" command to the monitor.  It can disassemble
    any range of memory ("disassemble c000:c010").  If labels have been
    defined, the disassembly will show them in the operands.

  - Added a new "assemble" command to the monitor.  It can assemble a 
    single instruction at an address ("assemble c000 jsr $ffd2").  
    Labels in the operands are also supported ("assemble c000 jsr charout").

  - Moved the character I/O area from $E000 to $F000 for compatibility with
    the EhBASIC binary saved from Michal Kowalski's Windows-based simulator.  
    In a future version of Py65, the I/O area will be configurable.

  - When running a program in the monitor, a read to $F004 will now do a
    non-blocking read from STDIN.  If no character is available, a null
    byte ($00) will be returned.

  - Fixed a bug where a CMP instruction could crash the simulator due to 
    an undefined variable.

  - EhBASIC 2.09 now runs in the simulator!

  - Documented all remaining monitor commands.  In the monitor, use the
    command "help command" for help on any command.

0.1 (2008-11-21)

  - First release.


src/py65/tests/test_memory.py
import unittest
import sys
import re
import os
from py65.memory import ObservableMemory

class ObservableMemoryTests(unittest.TestCase):

    # __setitem__

    def test___setitem__with_no_listeners_changes_memory(self):
        subject = self._make_subject()
        mem = ObservableMemory(subject=subject)
 
        mem[0xC000] = 0xAB
        self.assertEqual(0xAB, subject[0xC000])
 
    def test___setitem__ignores_subscribers_returning_none(self):
        subject = self._make_subject()
        mem = ObservableMemory(subject=subject)

        def write_subscriber_1(address, value):
            return None

        def write_subscriber_2(address, value):
            return None
        
        mem.subscribe_to_write([0xC000], write_subscriber_1)
        mem.subscribe_to_write([0xC000], write_subscriber_2)
        
        mem[0xC000] = 0xAB
        self.assertEqual(0xAB, subject[0xC000])

    def test___setitem__uses_result_of_last_subscriber(self):
        subject = self._make_subject()
        mem = ObservableMemory(subject=subject)
    
        def write_subscriber_1(address, value):
            return 0x01
    
        def write_subscriber_2(address, value):
            return 0x02
        
        mem.subscribe_to_write([0xC000], write_subscriber_1)
        mem.subscribe_to_write([0xC000], write_subscriber_2)
        
        mem[0xC000] = 0xAB
        self.assertEqual(0x02, subject[0xC000])
     
    # subscribe_to_read
    
    def test_subscribe_to_read_covers_all_addresses_in_range(self):
        subject = self._make_subject()
        mem = ObservableMemory(subject=subject)
    
        def read_subscriber(address, value):
            return 0xAB
        
        mem.subscribe_to_read(range(0xC000, 0xC001+1), read_subscriber)
    
        mem[0xC000] = 0xAB
        mem[0xC001] = 0xAB
        self.assertEqual(0xAB, subject[0xC001])
        self.assertEqual(0xAB, subject[0xC001])
     
    def test__subscribe_to_read_does_not_register_same_listener_twice(self):
        subject = self._make_subject()
        mem = ObservableMemory(subject=subject)
        
        calls = []
        def read_subscriber(address):
            calls.append('read_subscriber')
        
        mem.subscribe_to_read([0xC000], read_subscriber)
        mem.subscribe_to_read([0xC000], read_subscriber)
    
        value = mem[0xC000]
        self.assertEqual(['read_subscriber'], calls)
       
    # __getitem__
    
    def test___getitem__with_no_write_subscribers_changes_memory(self):
        subject = self._make_subject()
        mem = ObservableMemory(subject=subject)
     
        subject[0xC000] = 0xAB
        self.assertEqual(0xAB, mem[0xC000])
       
    def test___getitem__ignores_read_subscribers_returning_none(self):
        subject = self._make_subject()
        mem = ObservableMemory(subject=subject)
    
        def read_subscriber_1(address):
            return None
    
        def read_subscriber_2(address):
            return None
        
        mem.subscribe_to_read([0xC000], read_subscriber_1)
        mem.subscribe_to_read([0xC000], read_subscriber_2)
        
        mem[0xC000] = 0xAB
        self.assertEqual(0xAB, subject[0xC000])
        
    def test___getitem__calls_all_read_subscribers_uses_last_result(self):
        subject = self._make_subject()
        mem = ObservableMemory(subject=subject)
        
        calls = []
        def read_subscriber_1(address):
            calls.append('read_subscriber_1')
            return 0x01
    
        def read_subscriber_2(address):
            calls.append('read_subscriber_2')
            return 0x02
        
        mem.subscribe_to_read([0xC000], read_subscriber_1)
        mem.subscribe_to_read([0xC000], read_subscriber_2)
    
        subject[0xC000] = 0xAB
        self.assertEqual(0x02, mem[0xC000])

        expected_calls = ['read_subscriber_1', 'read_subscriber_2']
        self.assertEqual(expected_calls, calls)
     
    # __getattr__
    
    def test__getattr__proxies_subject(self):
        subject = self._make_subject()
        mem = ObservableMemory(subject=subject)
        self.assertEqual(subject.count, mem.count)
    
    # write
    
    def test_write_directly_writes_values_to_subject(self):
        subject = self._make_subject()
        mem = ObservableMemory(subject=subject)
    
        def write_subscriber(address, value):
            return 0xFF
        mem.subscribe_to_write([0xC000,0xC001], write_subscriber)
        
        mem.write(0xC000, [0x01, 0o02])
        self.assertEqual(0x01, subject[0xC000])
        self.assertEqual(0x02, subject[0xC001])

    # Test Helpers
    
    def _make_subject(self):    
        subject = 0x10000 * [0x00] 
        return subject

def test_suite():
    return unittest.findTestCases(sys.modules[__name__])

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')



src/py65/memory.py
from collections import defaultdict

class ObservableMemory:
    def __init__(self, subject=None):
        if subject is None:
            subject = 0x10000 * [0x00]
        self._subject = subject

        self._read_subscribers  = defaultdict(list)
        self._write_subscribers = defaultdict(list)        

    def __setitem__(self, address, value):
        if isinstance(address,slice):
            for n,v in zip(range(*address.indices(65536)),value):
                self[n] = v
            return
        callbacks = self._write_subscribers[address]
        for callback in callbacks:
            result = callback(address, value)
            if result is not None:
                value = result

        self._subject[address] = value
        
    def __getitem__(self, address):
        if isinstance(address,slice):
            return [self[n] for n in range(*address.indices(65536))]
        callbacks = self._read_subscribers[address]
        final_result = None

        for callback in callbacks:
            result = callback(address)
            if result is not None:
                final_result = result

        if final_result is None:
            return self._subject[address]
        else:
            return final_result
    
    def __getattr__(self, attribute):
        return getattr(self._subject, attribute)

    def subscribe_to_write(self, address_range, callback):
        for address in address_range:
            if not callback in self._write_subscribers[address]:
                self._write_subscribers[address].append(callback)

    def subscribe_to_read(self, address_range, callback): 
        for address in address_range:
            if not callback in self._read_subscribers[address]:
                self._read_subscribers[address].append(callback)

    def write(self, start_address, bytes):
        self._subject[start_address:start_address+len(bytes)] = bytes


src/py65/monitor.py
#!/usr/bin/env python -u

import cmd
import os
import re
import shlex
import sys
from asyncore import compact_traceback
from py65.devices.mpu6502 import MPU as NMOS6502
from py65.devices.mpu65c02 import MPU as CMOS65C02
from py65.disassembler import Disassembler
from py65.assembler import Assembler
from py65.utils.addressing import AddressParser
from py65.utils import console
from py65.utils.conversions import itoa
from py65.memory import ObservableMemory

class Monitor(cmd.Cmd):

    def __init__(self, mpu_type=NMOS6502, completekey='tab', stdin=None, stdout=None):
        self._reset(mpu_type)
        self._width = 78
        self._update_prompt()
        self._add_shortcuts()
        cmd.Cmd.__init__(self, completekey, stdin, stdout)

    def onecmd(self, line):
        line = self._preprocess_line(line)
        
        result = None
        try:
            result = cmd.Cmd.onecmd(self, line)
        except KeyboardInterrupt:
            self._output("Interrupt")
        except Exception as e:
            (file, fun, line), t, v, tbinfo = compact_traceback()
            error = 'Error: %s, %s: file: %s line: %s' % (t, v, file, line)
            self._output(error)

        self._update_prompt()
        return result

    def _reset(self, mpu_type):
        self._mpu = mpu_type()
        self._install_mpu_observers()
        self._address_parser = AddressParser()
        self._disassembler = Disassembler(self._mpu, self._address_parser)
        self._assembler = Assembler(self._mpu, self._address_parser)

    def _add_shortcuts(self):
        self._shortcuts = {'~':   'tilde',
                           '?':   'help',
                           'a':   'assemble',
                           'al':  'add_label',
                           'd':   'disassemble',
                           'dl':  'delete_label',
                           'f':   'fill',
                           '>':   'fill',
                           'g':   'goto',
                           'l':   'load',
                           'm':   'mem',
                           'r':   'registers',
                           'ret': 'return',
                           'rad': 'radix',
                           's':   'save',
                           'shl': 'show_labels',
                           'x':   'quit',
                           'z':   'step'}

    def _preprocess_line(self, line):
        # line comments
        quoted = False
        for pos, char in enumerate(line):
            if char in ('"', "'"):
                quoted = not quoted
            if (not quoted) and (char == ';'):
                line = line[:pos]
                break

        # whitespace & leading dots
        line = line.strip(' \t').lstrip('.')

        # special case for vice compatibility
        if line.startswith('~'):
          line = self._shortcuts['~'] + ' ' + line[1:]
      
        # command shortcuts
        for shortcut, command in self._shortcuts.items():
            if line == shortcut:
                line = command
                break

            pattern = '^%s\s+' % re.escape(shortcut)
            matches = re.match(pattern, line)
            if matches:
                start, end = matches.span()
                line = "%s %s" % (command, line[end:])
                break

        return line

    def _install_mpu_observers(self):
        def putc(address, value):
            self.stdout.write(chr(value))
            self.stdout.flush()

        def getc(address):
            char = console.getch_noblock(self.stdin)
            if char:
                byte = ord(char)
            else:
                byte = 0
            return byte

        m = ObservableMemory()
        #m.subscribe_to_write([0xF001], putc)
        #m.subscribe_to_read([0xF004], getc)
        
        self._mpu.memory = m

    def _update_prompt(self):
        self.prompt = "\n%s\n." % repr(self._mpu)
        
    def _output(self, stuff):
        if stuff is not None:
            self.stdout.write(stuff + "\n")

    def do_help(self, args):
        args = self._shortcuts.get(args.strip(), args)
        return cmd.Cmd.do_help(self, args)

    def help_version(self):
        self._output("version\t\tDisplay Py65 version information.")

    def do_version(self, args):
        self._output("\nPy65 Monitor")

    def help_help(self):
        self._output("help\t\tPrint a list of available actions.")
        self._output("help <action>\tPrint help for <action>.")

    def help_reset(self):
        self._output("reset\t\tReset the microprocessor")

    def do_reset(self, args):
        klass = self._mpu.__class__
        self._reset(mpu_type=klass)

    def do_mpu(self, args):                                         
        mpus = {'6502': NMOS6502, '65C02': CMOS65C02}
        
        def available_mpus():
            mpu_list = ', '.join(list(mpus.keys()))
            self._output("Available MPUs: %s" % mpu_list)            
        
        if args == '':                      
            self._output("Current MPU is %s" % self._mpu.name)
            available_mpus()
        else:
            requested = args.upper()
            new_mpu = mpus.get(requested, None)
            if new_mpu is None:
                self._output("Unknown MPU: %s" % args)
                available_mpus()
            else:
                self._reset(new_mpu)
                self._output("Reset with new MPU %s" % self._mpu.name)

    def help_mpu(self):
        self._output("mpu\t\tPrint available microprocessors.")
        self._output("mpu <type>\tSelect a new microprocessor.")
        
    def do_EOF(self, args):
        self._output('')
        return 1

    def help_EOF(self):
        self._output("To quit, type ^D or use the quit command.")

    def do_quit(self, args):
        return self.do_EOF(args)
    
    def help_quit(self):
        return self.help_EOF()

    def do_assemble(self, args):
        split = args.split(None, 1)
        if len(split) != 2:
            return self._interactive_assemble(args)
        
        start, statement = split
        try:
            start = self._address_parser.number(start)
        except KeyError:
            self._output("Bad label: %s" % start)
            return

        bytes = self._assembler.assemble(statement, start)
        if bytes is None:
            self._output("Assemble failed: %s" % statement)
        else:
            end = start + len(bytes)
            self._mpu.memory[start:end] = bytes
            self.do_disassemble('%04x:%04x' % (start, end))

    def help_assemble(self):
        self._output("assemble <address> <statement>")
        self._output("Assemble a statement at the address.")

    def _interactive_assemble(self, args):
        if args == '':
          start = self._mpu.pc
        else:
          try:
              start = self._address_parser.number(args)
          except KeyError:
              self._output("Bad label: %s" % start)
              return
          
        assembling = True

        while assembling:
          prompt = "\r$%04x            " % (start)
          line = console.line_input(prompt, 
                    stdin=self.stdin, stdout=self.stdout)

          if not line:
            self.stdout.write("\n")
            return

          # assemble into memory
          bytes = self._assembler.assemble(line)
          if bytes is None:
              self.stdout.write("\r$%04x  ???\n" % start)
              continue
          end = start + len(bytes)
          self._mpu.memory[start:end] = bytes  

          # print disassembly
          bytes, disasm = self._disassembler.instruction_at(start)
          disassembly = self._format_disassembly(start, bytes, disasm)
          self.stdout.write("\r" + (' ' * (len(prompt+line) + 5) ) + "\r")
          self.stdout.write(disassembly + "\n")

          start += bytes

    def do_disassemble(self, args):
        start, end = self._address_parser.range(args)
        if start == end:
            end += 1
        
        address = start
        while address < end:
            bytes, disasm = self._disassembler.instruction_at(address)
            self._output(self._format_disassembly(address, bytes, disasm))                
            address += bytes
    
    def _format_disassembly(self, address, bytes, disasm):
        mem = ''
        for byte in self._mpu.memory[address:address+bytes]:
            mem += '%02x ' % byte
        
        return "$%04x  %-10s%s" % (address, mem, disasm)

    def help_disassemble(self):
        self._output("disassemble <address_range>")
        self._output("Disassemble instructions in the address range.")

    def help_step(self):
        self._output("step")
        self._output("Single-step through instructions.")

    def do_step(self, args):
        self._mpu.step()
        self.do_disassemble('%04x' % self._mpu.pc)
    
    def help_return(self):
        self._output("return")
        self._output("Continues execution and returns to the monitor just")
        self._output("before the next RTS or RTI is executed.")
    
    def do_return(self, args):
        returns = [0x60, 0x40] # RTS, RTI
        self._run(stopcodes=returns)

    def help_goto(self):
        self._output("goto <address>")
        self._output("Change the PC to address and continue execution.")

    def do_goto(self, args):
        self._mpu.pc = self._address_parser.number(args)
        brks = [0x00] # BRK
        self._run(stopcodes=brks)
    
    def _run(self, stopcodes=[]):
        last_instruct = None
        while last_instruct not in stopcodes:
            self._mpu.step()
            last_instruct = self._mpu.memory[self._mpu.pc]
    
    def help_radix(self):
        self._output("radix [H|D|O|B]")
        self._output("Set the default radix to hex, decimal, octal, or binary.")
        self._output("With no argument, the current radix is printed.")
    
    def help_cycles(self):
        self._output("Display the total number of cycles executed.")
    
    def do_cycles(self, args):
        self._output(str(self._mpu.processorCycles))
    
    def do_radix(self, args):
        radixes = {'Hexadecimal': 16, 'Decimal': 10, 'Octal': 8, 'Binary': 2}

        if args != '':
            new = args[0].lower()
            changed = False
            for name, radix in radixes.items():
                if name[0].lower() == new:
                    self._address_parser.radix = radix
                    changed = True
            if not changed:
                self._output("Illegal radix: %s" % args)

        for name, radix in radixes.items():
            if self._address_parser.radix == radix:
                self._output("Default radix is %s" % name)

    def help_tilde(self):
        self._output("~ <number>")
        self._output("Display the specified number in decimal, hex, octal, and binary.")
    
    def do_tilde(self, args):   
        try:
            num = self._address_parser.number(args)
        except ValueError:
            self._output("Syntax error: %s" % args)
            return

        self._output("+%u" % num)
        self._output("$%02x" % num)
        self._output("%04o" % num)
        self._output(itoa(num, 2).zfill(8))
    
    def help_registers(self):
        self._output("registers[<reg_name> = <number> [, <reg_name> = <number>]*]")
        self._output("Assign respective registers.  With no parameters,")
        self._output("display register values.")
    
    def do_registers(self, args):
        if args == '':
            return
        
        pairs = re.findall('([^=,\s]*)=([^=,\s]*)', args)
        if pairs == []:
            return self._output("Syntax error: %s" % args)

        for register, value in pairs:
            if register not in ('pc', 'sp', 'a', 'x', 'y', 'p'):
                self._output("Invalid register: %s" % register)
            else:
                try:
                    intval = self._address_parser.number(value) & 0xFFFF
                    if len(register) == 1:
                        intval &= 0xFF
                    setattr(self._mpu, register, intval)
                except KeyError as why:
                    self._output(str(why))
    
    def help_cd(self, args):
        self._output("cd <directory>")
        self._output("Change the working directory.")
    
    def do_cd(self, args):
        try:
            os.chdir(args)
        except OSError as why:
            msg = "Cannot change directory: [%d] %s" % (why.errno, why.strerror)
            self._output(msg)
        self.do_pwd()

    def help_pwd(self):
        self._output("Show the current working directory.")
        
    def do_pwd(self, args=None):
        cwd = os.getcwd()
        self._output(cwd)

    def help_load(self):
        self._output("load \"filename\" <address>")
        self._output("Load the specified file into memory at the specified address.")
        self._output("Commodore-style load address bytes are ignored.")

    def do_load(self, args):
        split = shlex.split(args)
        if len(split) > 2:
            self._output("Syntax error: %s" % args)
            return

        filename = split[0]
        if len(split) == 2:
            start = self._address_parser.number(split[1])
        else:
            start = self._mpu.pc

        try:
            f = open(filename, 'rb')
            bytes = f.read()
            f.close()
        except (OSError, IOError) as why:
            msg = "Cannot load file: [%d] %s" % (why[0], why[1])
            self._output(msg)
            return


#        self._fill(start, start, list(map(ord, bytes)))
        self._fill(start, start, list(bytes))

    def do_save(self, args):
        split = shlex.split(args)
        if len(split) != 3:
            self._output("Syntax error: %s" % args)
            return
        
        filename = split[0]
        start = self._address_parser.number(split[1])
        end   = self._address_parser.number(split[2])
        
        bytes = bytearray(self._mpu.memory[start:end+1])
        try:
            f = open(filename, 'wb')
            f.write(bytes)
            f.close()
        except (OSError, IOError) as why:
            msg = "Cannot save file: [%d] %s" % (why.errno, why.strerror)
            self._output(msg)
            return
        
        self._output("Saved +%d bytes to %s" % (len(bytes), filename))    

    def help_save(self):
        self._output("save \"filename\" <start> <end>")
        self._output("Save the specified memory range to disk as a binary file.")
        self._output("Commodore-style load address bytes are not written.")
    
    def help_fill(self):
        self._output("fill <address_range> <data_list>")
        self._output("Fill memory in the specified address range with the data in")
        self._output("<data_list>.  If the size of the address range is greater")
        self._output("than the size of the data_list, the data_list is repeated.")
    
    def do_fill(self, args):
        split = shlex.split(args)
        if len(split) < 2:
            self._output("Syntax error: %s" % args)
            return

        start, end = self._address_parser.range(split[0])
        filler = list(map(self._address_parser.number, split[1:]))
        
        self._fill(start, end, filler)

    def _fill(self, start, end, filler):
        address = start
        length, index = len(filler), 0

        if start == end:
            end = start + length - 1
            if (end > 0xFFFF):
                end = 0xFFFF

        while address <= end:
            address &= 0xFFFF
            self._mpu.memory[address] = (filler[index] & 0xFF)
            index += 1
            if index == length:
                index = 0
            address += 1

        fmt = (end - start + 1, start, end)
        self._output("Wrote +%d bytes from $%04x to $%04x" % fmt)
    
    def help_mem(self):
        self._output("mem <address_range>")
        self._output("Display the contents of memory.")
    
    def do_mem(self, args):
        start, end = self._address_parser.range(args)

        line = "%04x:" % start
        for address in range(start, end+1):
            byte = self._mpu.memory[address]
            more = "  %02x" % byte                       
            
            exceeded = len(line) + len(more) > self._width
            if exceeded:
                self._output(line)
                line = "%04x:" % address
            line += more
        self._output(line)

    def help_add_label(self):
        self._output("add_label <address> <label>")
        self._output("Map a given address to a label.")

    def do_add_label(self, args):
        split = shlex.split(args)
        if len(split) != 2:
            self._output("Syntax error: %s" % args)
            return
        
        address = self._address_parser.number(split[0])    
        label   = split[1]

        self._address_parser.labels[label] = address

    def help_show_labels(self):
        self._output("show_labels")
        self._output("Display current label mappings.")

    def do_show_labels(self, args):
        values = list(self._address_parser.labels.values())
        keys = list(self._address_parser.labels.keys())
      
        byaddress = list(zip(values, keys))
        byaddress.sort()
        for address, label in byaddress:
            self._output("%04x: %s" % (address, label))

    def help_delete_label(self):
        self._output("delete_label <label>")
        self._output("Remove the specified label from the label tables.")

    def do_delete_label(self, args):
        if args == '':
            return self.help_delete_label()
        
        try:
            del self._address_parser.labels[args]
        except KeyError:
            pass

    def do_width(self, args):
        if args != '':
            try:
                new_width = int(args)
                if new_width >= 10:
                    self._width = new_width
                else:
                    self._output("Minimum terminal width is 10")
            except ValueError:
                self._output("Illegal width: %s" % args)

        self._output("Terminal width is %d" % self._width)

    def help_width(self):
        self._output("width <columns>")
        self._output("Set the width used by some commands to wrap output.")
        self._output("With no argument, the current width is printed.")

def main(args=None):
    c = Monitor()

    try:
        import readline
    except ImportError:
        pass

    try:
        c.onecmd('version')
        c.cmdloop()
    except KeyboardInterrupt:
        c._output('')
        pass

if __name__ == "__main__":
    main()
